group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/SimpleAssignor.java

```diff
@@ -72,36 +77,178 @@ private GroupAssignment assignHomogenous(
+
+        // Step 2: Round-robin assignment for unassigned partitions which do not have members already assigned in the current assignment.
+        Set<TargetPartition> assignedPartitions = new LinkedHashSet<>(newAssignment.keySet());
+        Map<Uuid, List<TargetPartition>> unassignedPartitions = new HashMap<>();
+        targetPartitions.forEach(targetPartition -> {
+            if (!assignedPartitions.contains(targetPartition) && !currentAssignment.containsKey(targetPartition))
+                unassignedPartitions.computeIfAbsent(targetPartition.topicId(), k -> new ArrayList<>()).add(targetPartition);
+        });
+
+        unassignedPartitions.keySet().forEach(unassignedTopic ->
+            roundRobinAssignment(topicToMemberSubscription.get(unassignedTopic), unassignedPartitions.get(unassignedTopic), newAssignment));
+
+        // Step 3: We combine current assignment and new assignment.
+        Map<String, Set<TargetPartition>> finalAssignment = new HashMap<>();
+
+        // When combining current assignment, we need to only consider the member topic subscription in current assignment
+        // which is being subscribed in the new assignment as well.
+        currentAssignment.forEach((targetPartition, members) -> members.forEach(member -> {
+            if (topicToMemberSubscription.getOrDefault(targetPartition.topicId(), Collections.emptySet()).contains(member))
+                finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition);
+        }));
+        newAssignment.forEach((targetPartition, members) -> members.forEach(member ->
+            finalAssignment.computeIfAbsent(member, k -> new HashSet<>()).add(targetPartition)));
+
+        return groupAssignment(finalAssignment, groupSpec.memberIds());
+    }
+
+    private GroupAssignment groupAssignment(
+        Map<String, Set<TargetPartition>> assignmentByMember,
+        Collection<String> allGroupMembers) {
+        Map<String, MemberAssignment> members = new HashMap<>();
+        for (Map.Entry<String, Set<TargetPartition>> entry : assignmentByMember.entrySet()) {
+            Map<Uuid, Set<Integer>> targetPartitions = new HashMap<>();
+            entry.getValue().forEach(targetPartition -> targetPartitions.computeIfAbsent(targetPartition.topicId(), k -> new HashSet<>()).add(targetPartition.partition()));
+            members.put(entry.getKey(), new MemberAssignmentImpl(targetPartitions));
+        }
+        allGroupMembers.forEach(member -> {
+            if (!members.containsKey(member))
+                members.put(member, new MemberAssignmentImpl(new HashMap<>()));
+        });
+
         return new GroupAssignment(members);
     }
 
-    private Map<Uuid, Set<Integer>> computeTargetPartitions(
+    // Visible for testing.
+    void memberHashAssignment(
+        List<TargetPartition> targetPartitions,
+        Collection<String> memberIds,
+        Map<TargetPartition, List<String>> assignment) {
```

Comment:
nit: As a general formatting point, indenting the arguments by 4 spaces, and then leaving the closing `) {` on the same line as the final argument means that the arguments and the following code statements are indented identically with no gap between them. Please reformat so the statements and the arguments are distinct. Personally, I would either place `) {` on a separate line, or indent all of the arguments at the end of the method name.